File: astrid/src/androidTest/java/com/todoroo/astrid/dao/TaskDaoTests.java
/**
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.dao;

import com.todoroo.andlib.data.Property;
import com.todoroo.andlib.data.TodorooCursor;
import com.todoroo.andlib.service.Autowired;
import com.todoroo.andlib.sql.Order;
import com.todoroo.andlib.sql.Query;
import com.todoroo.andlib.utility.DateUtilities;
import com.todoroo.astrid.dao.TaskDao.TaskCriteria;
import com.todoroo.astrid.data.Task;
import com.todoroo.astrid.test.DatabaseTestCase;

public class TaskDaoTests extends DatabaseTestCase {

    public static Property<?>[] IDS = new Property<?>[] { Task.ID };

    public static Property<?>[] TITLES = new Property<?>[] { Task.ID,
            Task.TITLE };

    @Autowired TaskDao taskDao;

    /**
     * Test basic task creation, fetch, and save
     */
    public void disabled_testTaskCreation() throws Exception {
        TodorooCursor<Task> cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(0, cursor.getCount());
        cursor.close();

        // create task "happy"
        Task task = new Task();
        task.setTitle("happy");
        assertTrue(taskDao.save(task));
        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(1, cursor.getCount());
        cursor.close();
        long happyId = task.getId();
        assertNotSame(Task.NO_ID, happyId);
        task = taskDao.fetch(happyId, TITLES);
        assertEquals("happy", task.getTitle());

        // create task "sad"
        task = new Task();
        task.setTitle("sad");
        assertTrue(taskDao.save(task));
        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(2, cursor.getCount());
        cursor.close();

        // rename sad to melancholy
        long sadId = task.getId();
        assertNotSame(Task.NO_ID, sadId);
        task.setTitle("melancholy");
        assertTrue(taskDao.save(task));
        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(2, cursor.getCount());
        cursor.close();

        // check state
        task = taskDao.fetch(happyId, TITLES);
        assertEquals("happy", task.getTitle());
        task = taskDao.fetch(sadId,TITLES);
        assertEquals("melancholy", task.getTitle());
    }

    /**
     * Test various task fetch conditions
     */
    public void disabled_testTaskConditions() throws Exception {
        // create normal task
        Task task = new Task();
        task.setTitle("normal");
        assertTrue(taskDao.save(task));

        // create blank task
        task = new Task();
        task.setTitle("");
        assertTrue(taskDao.save(task));

        // create hidden task
        task = new Task();
        task.setTitle("hidden");
        task.setHideUntil(DateUtilities.now() + 10000);
        assertTrue(taskDao.save(task));

        // create task with deadlines
        task = new Task();
        task.setTitle("deadlineInFuture");
        task.setDueDate(DateUtilities.now() + 10000);
        assertTrue(taskDao.save(task));

        task = new Task();
        task.setTitle("deadlineInPast");
        task.setDueDate(DateUtilities.now() - 10000);
        assertTrue(taskDao.save(task));

        // create completed task
        task = new Task();
        task.setTitle("completed");
        task.setCompletionDate(DateUtilities.now() - 10000);
        assertTrue(taskDao.save(task));

        // check has no name
        TodorooCursor<Task> cursor = taskDao.query(
                Query.select(TITLES).where(TaskCriteria.hasNoTitle()));
        assertEquals(1, cursor.getCount());
        cursor.moveToNext();
        assertEquals("", cursor.getString(1));
        cursor.close();

        // check has deadlines
        cursor = taskDao.query(Query.select(TITLES).where(
                TaskCriteria.hasDeadlines()).orderBy(Order.asc(Task.DUE_DATE)));
        assertEquals(2, cursor.getCount());
        cursor.moveToNext();
        assertEquals("deadlineInPast", cursor.getString(1));
        cursor.moveToNext();
        assertEquals("deadlineInFuture", cursor.getString(1));
        cursor.close();

        // check is active
        cursor = taskDao.query(Query.select(TITLES).where(TaskCriteria.
                isActive()));
        assertEquals(5, cursor.getCount());
        cursor.close();

        // check due before / after
        cursor = taskDao.query(Query.select(TITLES).where(TaskCriteria.
                dueBeforeNow()));
        cursor.moveToNext();
        assertEquals(1, cursor.getCount());
        cursor.close();
        cursor = taskDao.query(Query.select(TITLES).where(TaskCriteria.
                dueAfterNow()));
        assertEquals(1, cursor.getCount());
        cursor.close();

        // check completed before
        cursor = taskDao.query(Query.select(TITLES).where(TaskCriteria.
                completed()));
        assertEquals(1, cursor.getCount());
        cursor.close();

        // check is visible
        cursor = taskDao.query(Query.select(TITLES).where(TaskCriteria.
                isVisible()));
        assertEquals(5, cursor.getCount());
        cursor.close();
    }

    /**
     * Test task deletion
     */
    public void disabled_testTDeletion() throws Exception {
        TodorooCursor<Task> cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(0, cursor.getCount());
        cursor.close();

        // create task "happy"
        Task task = new Task();
        task.setTitle("happy");
        assertTrue(taskDao.save(task));
        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(1, cursor.getCount());
        cursor.close();

        // delete
        long happyId = task.getId();
        assertTrue(taskDao.delete(happyId));
        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(0, cursor.getCount());
        cursor.close();
    }

    /**
     * Test save without prior create doesn't work
     */
    public void disabled_testSaveWithoutCreate() throws Exception {
        TodorooCursor<Task> cursor;

        // try to save task "happy"
        Task task = new Task();
        task.setTitle("happy");
        task.setID(1L);

        assertFalse(taskDao.save(task));

        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(0, cursor.getCount());
        cursor.close();
    }

    /**
     * Test passing invalid task indices to various things
     */
    public void disabled_testInvalidIndex() throws Exception {
        TodorooCursor<Task> cursor;

        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(0, cursor.getCount());
        cursor.close();

        assertNull(taskDao.fetch(1, IDS));

        assertFalse(taskDao.delete(1));

        // make sure db still works
        cursor = taskDao.query(
                Query.select(IDS));
        assertEquals(0, cursor.getCount());
        cursor.close();
    }

    // TODO check eventing
}

File: app/src/androidTest/java/org/tasks/data/TaskDaoTests.kt
/*
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package org.tasks.data

import com.natpryce.makeiteasy.MakeItEasy.with
import org.tasks.data.entity.Task
import com.todoroo.astrid.service.TaskDeleter
import dagger.hilt.android.testing.HiltAndroidTest
import dagger.hilt.android.testing.UninstallModules
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import org.tasks.data.dao.TaskDao
import org.tasks.injection.InjectingTestCase
import org.tasks.injection.ProductionModule
import org.tasks.makers.TaskMaker.PARENT
import org.tasks.makers.TaskMaker.newTask
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import javax.inject.Inject

@UninstallModules(ProductionModule::class)
@HiltAndroidTest
class TaskDaoTests : InjectingTestCase() {

    @Inject lateinit var taskDao: TaskDao
    @Inject lateinit var taskDeleter: TaskDeleter

    /** Test various task fetch conditions  */
    @Test
    fun testTaskConditions() = runBlocking {
        // create normal task
        var task = Task()
        task.title = "normal"
        taskDao.createNew(task)

        // create blank task
        task = Task()
        task.title = ""
        taskDao.createNew(task)

        // create hidden task
        task = Task()
        task.title = "hidden"
        task.hideUntil = currentTimeMillis() + 10000
        taskDao.createNew(task)

        // create task with deadlines
        task = Task()
        task.title = "deadlineInFuture"
        task.dueDate = currentTimeMillis() + 10000
        taskDao.createNew(task)
        task = Task()
        task.title = "deadlineInPast"
        task.dueDate = currentTimeMillis() - 10000
        taskDao.createNew(task)

        // create completed task
        task = Task()
        task.title = "completed"
        task.completionDate = currentTimeMillis() - 10000
        taskDao.createNew(task)

        // check is active
        assertEquals(5, taskDao.getActiveTasks().size)

        // check is visible
        assertEquals(5, taskDao.getActiveTasks().size)
    }

    /** Test task deletion  */
    @Test
    fun testTDeletion() = runBlocking {
        assertEquals(0, taskDao.getAll().size)

        // create task "happy"
        val task = Task()
        task.title = "happy"
        taskDao.createNew(task)
        assertEquals(1, taskDao.getAll().size)

        // delete
        taskDeleter.delete(task)
        assertEquals(0, taskDao.getAll().size)
    }

    /** Test passing invalid task indices to various things  */
    @Test
    fun testInvalidIndex() = runBlocking {
        assertEquals(0, taskDao.getAll().size)
        assertNull(taskDao.fetch(1))
        taskDeleter.delete(listOf(1L))

        // make sure db still works
        assertEquals(0, taskDao.getAll().size)
    }

    @Test
    fun findChildrenInList() = runBlocking {
        val parent = taskDao.createNew(newTask())
        val child = taskDao.createNew(newTask(with(PARENT, parent)))
        assertEquals(listOf(child), taskDao.getChildren(listOf(parent, child)))
    }

    @Test
    fun findRecursiveChildrenInList() = runBlocking {
        val parent = taskDao.createNew(newTask())
        val child = taskDao.createNew(newTask(with(PARENT, parent)))
        val grandchild = taskDao.createNew(newTask(with(PARENT, child)))
        assertEquals(
                listOf(child, grandchild, grandchild),
                taskDao.getChildren(listOf(parent, child, grandchild)))
    }

    @Test
    fun findRecursiveChildrenInListAfterSkippingParent() = runBlocking {
        val parent = taskDao.createNew(newTask())
        val child = taskDao.createNew(newTask(with(PARENT, parent)))
        val grandchild = taskDao.createNew(newTask(with(PARENT, child)))
        assertEquals(listOf(child, grandchild), taskDao.getChildren(listOf(parent, grandchild)))
    }

    @Test
    fun dontSetParentToSelf() = runBlocking {
        val parent = taskDao.createNew(newTask())
        val child = taskDao.createNew(newTask())

        taskDao.setParent(parent, listOf(parent, child))

        assertEquals(0, taskDao.fetch(parent)!!.parent)
        assertEquals(parent, taskDao.fetch(child)!!.parent)
    }
}